File: Tustler/Helpers/ModuleResolver.cs
using CloudWeaver.AWS;
using CloudWeaver.Types;
using Microsoft.FSharp.Collections;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Tustler.Helpers
{
    /// <summary>
    /// Determines which deserialization function to call on which module
    /// </summary>
    /// <remarks>
    /// The standard module (StandardShareIntraModule) includes a SaveFlags argument type which also requires resolving
    /// in that a flag item may be defined in any module e.g. standard vs AWS flag items
    /// </remarks>
    public class ModuleResolver
    {
        private readonly Dictionary<string, Func<string, Dictionary<string, ISaveFlagSet>, Dictionary<string, ISaveFlagSet>>> _flagSetLookup;

        /// <summary>
        /// Determine which module to call for seserialization
        /// </summary>
        /// <param name="moduleTag">The name of the module (the type as a string)</param>
        /// <returns></returns>
        public static Func<string, string, IShareIntraModule> ModuleLookup(string moduleTag)
        {
            static Func<string, string, IShareIntraModule> GetStandardResolver()
            {
                var flagSetLookup = new Dictionary<string, Func<string, Dictionary<string, ISaveFlagSet>, Dictionary<string, ISaveFlagSet>>>()
                    {
                        { "StandardFlag", FoldInStandardValue },
                        { "AWSFlag", FoldInAWSValue }
                    };
                return new ModuleResolver(flagSetLookup).Deserialize;
            }

            return moduleTag switch
            {
                "StandardShareIntraModule" => GetStandardResolver(),
                "AWSShareIntraModule" => AWSShareIntraModule.Deserialize,
                _ => throw new ArgumentException($"Unexpected module tag ({moduleTag}) in ModuleLookup")
            };
        }

        public ModuleResolver(Dictionary<string, Func<string, Dictionary<string, ISaveFlagSet>, Dictionary<string, ISaveFlagSet>>> flagSetLookup)
        {
            _flagSetLookup = flagSetLookup;
        }

        public IShareIntraModule Deserialize(string propertyName, string jsonString)
        {
            return StandardShareIntraModule.Deserialize(propertyName, jsonString, _flagSetLookup);
        }

        static Dictionary<string, ISaveFlagSet> FoldInStandardValue(string serializedFlagItem, Dictionary<string, ISaveFlagSet> source)
        {
            if (StandardFlagItem.GetNames().Contains(serializedFlagItem))
            {
                var flagItem = StandardFlagItem.Create(serializedFlagItem);
                var standardFlag = new StandardFlag(flagItem);

                if (!source.ContainsKey("StandardFlagSet"))
                {
                    source.Add("StandardFlagSet", new StandardFlagSet());
                }

                var standardFlagSet = source["StandardFlagSet"] as StandardFlagSet;
                standardFlagSet.SetFlag(standardFlag);
            }

            return source;
        }

        static Dictionary<string, ISaveFlagSet> FoldInAWSValue(string serializedFlagItem, Dictionary<string, ISaveFlagSet> source)
        {
            if (AWSFlagItem.GetNames().Contains(serializedFlagItem))
            {
                var flagItem = AWSFlagItem.Create(serializedFlagItem);
                var awsFlag = new AWSFlag(flagItem);

                if (!source.ContainsKey("AWSFlagSet"))
                {
                    source.Add("AWSFlagSet", new AWSFlagSet());
                }

                var awsFlagSet = source["AWSFlagSet"] as AWSFlagSet;
                awsFlagSet.SetFlag(awsFlag);
            }

            return source;
        }
    }
}

File: CloudWeaver.Agent/ModuleResolver.fs
namespace CloudWeaver

open System.Collections.Generic
open System
open CloudWeaver.Types
open Converters
open TustlerServicesLib
open System.Text.Json


/// <summary>
/// Determines which deserialization function to call on which module
/// </summary>
/// <remarks>
/// The standard module (StandardShareIntraModule) includes a SaveFlags argument type which also requires resolving
/// in that a flag item may be defined in any module e.g. standard vs AWS flag items
/// </remarks>
type ModuleResolver (serializerOptions, flagSetLookup: Dictionary<string, Func<string, Dictionary<string, ISaveFlagSet>, Dictionary<string, ISaveFlagSet>>>) =

    /// Get the deserializer for Standard modules (wraps the flag module resolver)
    member this.Deserialize with get () =
        Func<_, _, _>(fun propertyName jsonString -> StandardShareIntraModule.Deserialize propertyName jsonString serializerOptions flagSetLookup :> IShareIntraModule)

    ///// Deserialize and add a new Standard flag itme
    //static member FoldInStandardValue (serializedFlagItem: string) (source: Dictionary<string, ISaveFlagSet>) =

    //    if StandardFlagItem.GetNames() |> Seq.contains(serializedFlagItem) then
    //        let flagItem = StandardFlagItem.Create(serializedFlagItem)
    //        let standardFlag = StandardFlag(flagItem)

    //        let standardFlagSet =
    //            if source.ContainsKey("StandardFlagSet") then
    //                (source.["StandardFlagSet"]) :?> StandardFlagSet
    //            else
    //                let flagSet = new StandardFlagSet()
    //                source.Add("StandardFlagSet", flagSet)
    //                flagSet
            
    //        standardFlagSet.SetFlag(standardFlag)

    //    source

    ///// Deserialize and add a new AWS flag itme
    //static member FoldInAWSValue (serializedFlagItem: string) (source: Dictionary<string, ISaveFlagSet>): Dictionary<string, ISaveFlagSet> =

    //    if AWSFlagItem.GetNames() |> Seq.contains(serializedFlagItem) then
    //        let flagItem = AWSFlagItem.Create(serializedFlagItem)
    //        let awsFlag = AWSFlag(flagItem)
            
    //        let awsFlagSet =
    //            if source.ContainsKey("AWSFlagSet") then
    //                (source.["AWSFlagSet"]) :?> AWSFlagSet
    //            else
    //                let flagSet = new AWSFlagSet()
    //                source.Add("AWSFlagSet", flagSet)
    //                flagSet

    //        awsFlagSet.SetFlag(awsFlag)

    //    source

    /// <summary>
    /// Determine which module to call for deserialization
    /// </summary>
    /// <param name="moduleTag">The name of the module (here using the module type as a string)</param>
    /// <returns>A deserialization function appropriate for the module</returns>
    static member ModuleLookup (moduleTag: string) =

        //let ToFunc flagModuleTag (typeResolver: TypeResolver) = 
        //    let flagResolver =
        //        match flagModuleTag with
        //        | "StandardFlag" -> ModuleResolver.FoldInStandardValue
        //        | "AWSFlag" -> ModuleResolver.FoldInAWSValue
        //        | _ -> invalidArg "flagModuleTag" (sprintf "Unexpected flag module tag (%s) in ModuleLookup" flagModuleTag)

        //    Func<_, _, _>(fun (serializedFlagItem: string) (source: Dictionary<string, ISaveFlagSet>) -> flagResolver serializedFlagItem source)

        let toFunc flagModuleTag (typeResolver: TypeResolver) = 
            Func<_, _, _>(fun (serializedFlagItem: string) (source: Dictionary<string, ISaveFlagSet>) -> typeResolver.AddFlag(flagModuleTag, serializedFlagItem, source))

        let getStandardResolver typeResolver serializerOptions =
            let flagSetLookup =
                let pairs = seq {
                    KeyValuePair( "StandardFlag", toFunc "StandardFlag" typeResolver );
                    KeyValuePair( "AWSFlag", toFunc "AWSFlag" typeResolver );
                }
                Dictionary<string, (Func<string, Dictionary<string, ISaveFlagSet>, Dictionary<string, ISaveFlagSet>>)>(pairs)
            ModuleResolver(serializerOptions, flagSetLookup).Deserialize

        let getModuleResolver (typeResolver: TypeResolver) typeName serializerOptions =
            let deserialize = typeResolver.ResolveStaticCall(typeName, "Deserialize") :?> Func<string, string, JsonSerializerOptions, IShareIntraModule>
            Func<_, _, _>(fun propertyName jsonString -> deserialize.Invoke(propertyName, jsonString, serializerOptions))

        let typeResolver = TypeResolver.Create() |> Async.AwaitTask |> Async.RunSynchronously
        
        let serializerOptions = CreateSerializerOptions(typeResolver)
        serializerOptions.Converters.Add(SentenceChunkerConverter())

        /// Get the deserializer for each module (wraps the flag module resolver for the Standard module)
        match moduleTag with
        | "StandardShareIntraModule" -> getStandardResolver typeResolver serializerOptions
        | "AWSShareIntraModule" -> getModuleResolver typeResolver "CloudWeaver.AWS.AWSShareIntraModule" serializerOptions
        | "AVShareIntraModule" -> getModuleResolver typeResolver "CloudWeaver.MediaServices.AVShareIntraModule" serializerOptions
        | _ -> invalidArg "moduleTag" (sprintf "Unexpected module tag (%s) in ModuleLookup" moduleTag)

